src/_core/usdPrices.tsx

```diff
@@ -0,0 +1,71 @@
+import {
+  createContext,
+  ReactNode,
+  useCallback,
+  useContext,
+  useEffect,
+  useState,
+} from 'react';
+import invariant from 'tiny-invariant';
+
+import { usePopupSelector } from '../popup/store/react';
+import Background from '../ui/services/Background';
+import { startPolling } from './polling';
+import { useDebouncedValue } from './useDebouncedValue';
+
+const UsdPricesContext = createContext<
+  null | ((assetIds: string[]) => undefined | (() => void))
+>(null);
+
+export function UsdPricesProvider({ children }: { children: ReactNode }) {
+  const [observedAssetIds, setObservedAssetIds] = useState<string[][]>([]);
+  const observedAssetIdsDebounced = useDebouncedValue(observedAssetIds, 100);
+
+  useEffect(() => {
+    if (!observedAssetIdsDebounced.length) {
+      return;
+    }
+
+    return startPolling(5000, () =>
+      Background.updateUsdPricesByAssetIds([
+        ...new Set(observedAssetIdsDebounced.flat()),
+      ])
+    );
+  }, [observedAssetIdsDebounced]);
+
+  const observe = useCallback((assetIds: string[]) => {
+    if (!assetIds.length) {
+      return;
+    }
+
+    setObservedAssetIds(ids => [...ids, assetIds]);
+
+    return () => {
+      setObservedAssetIds(ids => {
+        ids.splice(ids.indexOf(assetIds), 1);
+
+        return ids;
+      });
+    };
+  }, []);
+
+  return (
+    <UsdPricesContext.Provider value={observe}>
+      {children}
+    </UsdPricesContext.Provider>
+  );
+}
+
+export function useUsdPrices(assetIds: string[]) {
+  const observeUsdPricesByAssetIds = useContext(UsdPricesContext);
+  invariant(observeUsdPricesByAssetIds);
+
+  const usdPrices = usePopupSelector(state => state.usdPrices);
+
+  useEffect(
+    () => observeUsdPricesByAssetIds(assetIds),
+    [observeUsdPricesByAssetIds, assetIds]
+  );
+
+  return usdPrices;
```

Comment:
May be filter returned object to only have `assetIds` keys? Otherwise you can write `useUsdPrices([])` and get all the values, except they won't be kept fresh. Which in turn can be a debugging nightmare.

And probably wrap it into a `useMemo` then.